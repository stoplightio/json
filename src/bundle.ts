import { cloneDeep, get, has, set } from 'lodash';

import { hasRef } from './hasRef';
import { isLocalRef } from './isLocalRef';
import { pathToPointer } from './pathToPointer';
import { pointerToPath } from './pointerToPath';
import { traverse } from './traverse';

export const BUNDLE_ROOT = '__bundled__';

export const bundleTarget = <T = unknown>({ document, path }: { document: T; path: string }, cur?: unknown) =>
  _bundle(cloneDeep(document), path, cur);

const _bundle = (
  document: unknown,
  path: string,
  cur?: unknown,
  bundledRefInventory: any = {},
  bundledObj: any = {},
) => {
  const objectToBundle = get(document, pointerToPath(path));

  traverse(cur ? cur : objectToBundle, ({ parent }) => {
    if (hasRef(parent) && isLocalRef(parent.$ref)) {
      if (bundledRefInventory[parent.$ref]) {
        parent.$ref = bundledRefInventory[parent.$ref];

        // no need to continue, this $ref has already been bundled
        return;
      }

      const _path = pointerToPath(parent.$ref);
      const inventoryPath = [BUNDLE_ROOT, ..._path];
      const inventoryRef = pathToPointer(inventoryPath);

      const bundled$Ref = get(document, _path);
      if (bundled$Ref) {
        const pathProcessed = [];

        bundledRefInventory[parent.$ref] = inventoryRef;
        parent.$ref = inventoryRef;

        // make sure arrays and object decisions are preserved when copying over the portion of the tree
        for (const key of _path) {
          pathProcessed.push(key);

          const inventoryPathProcessed = [BUNDLE_ROOT, ...pathProcessed];
          if (has(bundledObj, inventoryPathProcessed)) continue;

          const target = get(document, pathProcessed);
          if (Array.isArray(target)) {
            set(bundledObj, inventoryPathProcessed, []);
          } else if (typeof target === 'object') {
            set(bundledObj, inventoryPathProcessed, {});
          }
        }

        set(bundledObj, inventoryPath, bundled$Ref);
        _bundle(document, path, bundled$Ref, bundledRefInventory, bundledObj);
      }
    }
  });

  set(objectToBundle, BUNDLE_ROOT, bundledObj[BUNDLE_ROOT]);

  return objectToBundle;
};
