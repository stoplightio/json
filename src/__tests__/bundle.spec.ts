import { cloneDeep } from 'lodash';

import { bundleTarget } from '../bundle';

describe('bundleTargetPath()', () => {
  it('should work', () => {
    const document = {
      definitions: {
        user: {
          id: 'foo',
          address: {
            $ref: '#/definitions/address',
          },
        },
        address: {
          street: 'foo',
          user: {
            $ref: '#/definitions/user',
          },
        },
        card: {
          zip: '20815',
        },
      },
      __target__: {
        entity: {
          $ref: '#/definitions/user',
        },
      },
    };

    const clone = cloneDeep(document);

    const result = bundleTarget({
      document: clone,
      path: '#/__target__',
    });

    // Do not mutate document
    expect(clone).toEqual(document);

    expect(result).toEqual({
      entity: {
        $ref: '#/definitions/user',
      },
      definitions: {
        user: {
          id: 'foo',
          address: {
            $ref: '#/definitions/address',
          },
        },
        address: {
          street: 'foo',
          user: {
            $ref: '#/definitions/user',
          },
        },
      },
    });
  });

  it('should not throw erorr', () => {
    const document = {
      definitions: {
        user: {
          id: 'foo',
          address: {
            $ref: '#/definitions/address',
          },
        },
        address: {
          street: 'foo',
          user: {
            $ref: '#/definitions/user',
          },
        },
      },
      __target__: {
        entity: {
          $ref: '#/definitions/user',
        },
        entity2: {
          $ref: './path/to/pet.json',
        },
      },
    };

    const clone = cloneDeep(document);

    const result = bundleTarget({
      document: clone,
      path: '#/__target__',
    });

    // Do not mutate document
    expect(clone).toEqual(document);

    expect(result).toEqual({
      entity: {
        $ref: '#/definitions/user',
      },
      entity2: {
        $ref: './path/to/pet.json',
      },
      definitions: {
        user: {
          id: 'foo',
          address: {
            $ref: '#/definitions/address',
          },
        },
        address: {
          street: 'foo',
          user: {
            $ref: '#/definitions/user',
          },
        },
      },
    });
  });

  it('should mirror original source decision re arrays or objects', () => {
    const document = {
      parameters: [
        {},
        {
          schema: {
            name: 'param',
          },
        },
      ],
      responses: {
        '200': {
          other: 'foo',
          schema: {
            title: 'OK',
            parameter: {
              $ref: '#/parameters/1/schema',
            },
          },
        },
      },
      __target__: {
        entity: {
          $ref: '#/responses/200/schema',
        },
      },
    };

    const clone = cloneDeep(document);

    const result = bundleTarget({
      document: clone,
      path: '#/__target__',
    });

    // Do not mutate document
    expect(clone).toEqual(document);

    expect(result).toEqual({
      entity: {
        $ref: '#/responses/200/schema',
      },
      parameters: [
        undefined,
        {
          schema: {
            name: 'param',
          },
        },
      ],
      responses: {
        '200': {
          schema: {
            title: 'OK',
            parameter: {
              $ref: '#/parameters/1/schema',
            },
          },
        },
      },
    });
  });
});
