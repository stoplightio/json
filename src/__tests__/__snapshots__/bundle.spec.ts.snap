// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundleTargetPath() should handle deep circular refs 1`] = `
Object {
  "__bundled__": Object {
    "components": Object {
      "schemas": Object {
        "GeographicalCoordinate": Object {
          "type": "object",
        },
        "Location": [Circular],
      },
    },
  },
  "properties": Object {
    "PhysicalGeographicalCoordinate": Object {
      "$ref": "#/__bundled__/components/schemas/GeographicalCoordinate",
    },
    "RelatedLocation": Object {
      "$ref": "#/__bundled__/components/schemas/Location",
    },
  },
  "type": "object",
}
`;
